### define the following in short:

## Association Analysis:

## Support:

### Confidence:

### The APRIORI PRINCIPLE:

### Candidate Set:

### Min support threshold:

### Frequent Item Set:


### Apriori Algorithm:
### $ Notation -->  C_{k} = Candidate \;itemset \;of \;size \; k \; and\; L_{k} =  frequent \;itemset\; of\; size\; k $
### $  C_{k+1} =\; candidates \;generated \;from \;L_{k}$

### $ L_{k +1} = candidates \;in \; C_{k+1} \;satisfying\; minsup $

1. Read the transaction database and get support for each itemset, compare the support with minimum support to generate frequent itemset at level 1.
2.  Use join to generate a set of candidate k-item sets of length $ K+1 ( C_{k+1} ) $at next level.
3. Generate frequent item sets sets of length $ K+1 (L_{k +1})$at next level using minimum support.
     In this step
        3.1 scan original database  to count support for k+1 candidates
        3.2 prune candidates below minsup
4.  Repeat step 2 and 3 until no frequent item sets can be generated.
5. Generate rules form frequent itemsets from level 2 onwards using minimum confidence.


# Implement apriori algorithm for market basket analysis

In [ ]:
 ! pip install apyori  # code  to install and use apriori algorithm model

In [ ]:
import numpy as np
import pandas as pd
from apyori import apriori

In [ ]:
data_frame = pd.read_csv('C:/Users/Administrator/apriori/market_basket.csv', header =None)
data_frame.head()

0      1      2       3     4      5
0  Wine  Chips  Bread  Butter  Milk  Apple
1  Wine    NaN  Bread  Butter  Milk    NaN
2   NaN    NaN  Bread  Butter  Milk    NaN
3   NaN  Chips    NaN  Butter   NaN  Apple
4  Wine  Chips  Bread  Butter  Milk  Apple

In [ ]:
data_frame.shape

(22, 6)

# Convert Pandas dataframe into nested lists  

In [ ]:
lsts = []
for i in range (0,22):
    lsts.append([str(data_frame.values[i,j]) for j in range (0,6)])

print (lsts)

[['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'], ['Wine', 'nan', 'Bread', 'Butter', 'Milk', 'nan'], ['nan', 'nan', 'Bread', 'Butter', 'Milk', 'nan'], ['nan', 'Chips', 'nan', 'Butter', 'nan', 'Apple'], ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'], ['Wine', 'Chips', 'nan', 'nan', 'Milk', 'nan'], ['Wine', 'Chips', 'Bread', 'Butter', 'nan', 'Apple'], ['Wine', 'Chips', 'nan', 'nan', 'Milk', 'nan'], ['Wine', 'nan', 'Bread', 'nan', 'nan', 'Apple'], ['nan', 'nan', 'Bread', 'Butter', 'Milk', 'nan'], ['Wine', 'Chips', 'Bread', 'Butter', 'nan', 'Apple'], ['Wine', 'nan', 'nan', 'Butter', 'Milk', 'Apple'], ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'], ['Wine', 'nan', 'Bread', 'nan', 'Milk', 'nan'], ['Wine', 'nan', 'Bread', 'Butter', 'Milk', 'Apple'], ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'], ['nan', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'], ['nan', 'Chips', 'nan', 'Butter', 'Milk', 'Apple'], ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'], ['Wine', 'n

## Goal of Association Rule Mining and its Application to any buisness

### When you apply Association Rule Mining on a given set of transactions T your goal will be to find all rules with:

### 1. Support greater than or equal to min_support
### 2. Confidence greater than or equal to min_confidence

### One of the algorithm for Association Rule Mining implemented here is APRIORI



### Support - defines the popularity of an item

### Confidence - it tells the possibility of how often items X and Y occur together, given the number of times X has occurred

### Strong rules  - If a rule A=>B [Support, Confidence] satisfies min_sup and min_confidence then it is a strong rule.

###  Lift gives the correlation between A and B in the rule A=>B. Correlation shows how one item-set A effects the item-set B.
 ### Lift(A=>B)=   ( Support ) / (Supp(A)Supp(B))


### If the lift is > 1, then A and B are dependent on each other, and the degree of which is given by lift value.

###  Lift indicates certainty of a rule. How much sale of X has increased when Y is sold?

### Lift(X=>Y) = Confidence(X, Y)/Support(Y)
## Example: X => Y [support = 60%, confidence = 90%]
## 60% of all the transactions show that X and Y have been bought together.

## 90% of the customers who bought X also bought Y.



# Make APRIORI MODEL  for RULE GENERATION

In [ ]:
asscsn_rules = apriori(lsts, min_support =0.50, min_confidence = 0.7, min_lift = 1.2, min_length = 2)
asscsn_results = list(asscsn_rules)

In [ ]:
print(asscsn_results)

[RelationRecord(items=frozenset({'Wine', 'Apple', 'Bread'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Apple'}), items_add=frozenset({'Wine', 'Bread'}), confidence=0.7333333333333334, lift=1.241025641025641), OrderedStatistic(items_base=frozenset({'Apple', 'Bread'}), items_add=frozenset({'Wine'}), confidence=0.9166666666666667, lift=1.2604166666666667), OrderedStatistic(items_base=frozenset({'Wine', 'Apple'}), items_add=frozenset({'Bread'}), confidence=0.9166666666666667, lift=1.2604166666666667), OrderedStatistic(items_base=frozenset({'Wine', 'Bread'}), items_add=frozenset({'Apple'}), confidence=0.8461538461538461, lift=1.241025641025641)])]


# Result Interpretation -  Market basket analysis

##  Consumer behavior insights from the  given dataset and its applicaiton as interpreted below-

### Frequent Item set from the market basket dataset analysis is

### {'Wine', 'Apple', 'Bread'}, support=0.5, i.e these items are bought together 50% of the time  on overall transactions



#### Strng  R1- {'Apple'}-->{'Bread', 'Wine'}, confidence=0.7333333333333334, lift=1.241025641025641



# Result interpretation for  {'Apple'}-->{'Bread', 'Wine'}

###### i.e. confidence=0.7333333333333334 means 73.33% of times consumers who bought {'APPLE' } also bought { 'BREAD, WINE' }


#####  lift=1.241025641025641  means  {'Apple'} --> {'Bread', 'Wine'} are highly corelated and
#### Lift infers us "{'Bread', 'Wine'} is 1.24 times more likely to be bought by the customers who buy {'Apple'}

### the more greater than 1 the more correlation between them

#### Strng  R2- {'Apple', 'Bread'} -->{'Wine'}, confidence=0.9166666666666667, lift=1.2604166666666667

#### Strng  R3 - {'Apple', 'Wine'}-->{'Bread'}, confidence=0.9166666666666667, lift=1.2604166666666667

#### Strng  R4 -  {'Bread', 'Wine'}-->{'Apple'}, confidence=0.8461538461538461, lift=1.241025641025641)
